### ISSUE TITLE
Single object numpy arrays cause crash

### ISSUE THREAD

**Issue by takhs91** [ijl/orjson#107]
Hi,
congratulations on making the fastest python json lib.

There seem to be an issue with those edge case numpy arrays
```
>>> orjson.dumps({'a' : np.array(0)}, option=orjson.OPT_SERIALIZE_NUMPY)                                                                                                                                                                 
thread '<unnamed>' panicked at 'index out of bounds: the len is 0 but the index is 18446744073709551615', src/array.rs:155:9
stack backtrace:
   0: <unknown>
   1: <unknown>
   2: <unknown>
   3: <unknown>
   4: <unknown>
   5: <unknown>
   6: <unknown>
   7: <unknown>
   8: <unknown>
   9: <unknown>
  10: <unknown>
  11: dumps
  12: _PyMethodDef_RawFastCallKeywords
             at Objects/call.c:690
  13: _PyCFunction_FastCallKeywords
             at Objects/call.c:730
  14: call_function
             at Python/ceval.c:4568
  15: _PyEval_EvalFrameDefault
             at Python/ceval.c:3139
  16: PyEval_EvalFrameEx
             at Python/ceval.c:547
  17: _PyEval_EvalCodeWithName
             at Python/ceval.c:3930
  18: PyEval_EvalCodeEx
             at Python/ceval.c:3959
  19: PyEval_EvalCode
             at Python/ceval.c:524
  20: builtin_exec_impl
             at Python/bltinmodule.c:1079
  21: builtin_exec
             at Python/clinic/bltinmodule.c.h:283
  22: _PyMethodDef_RawFastCallKeywords
             at Objects/call.c:651
  23: _PyCFunction_FastCallKeywords
             at Objects/call.c:730
  24: call_function
             at Python/ceval.c:4568
  25: _PyEval_EvalFrameDefault
             at Python/ceval.c:3124
  26: gen_send_ex
             at Objects/genobject.c:221
  27: _PyGen_Send
             at Objects/genobject.c:291
  28: _PyEval_EvalFrameDefault
             at Python/ceval.c:1787
  29: gen_send_ex
             at Objects/genobject.c:221
  30: _PyGen_Send
             at Objects/genobject.c:291
  31: _PyEval_EvalFrameDefault
             at Python/ceval.c:1787
  32: gen_send_ex
             at Objects/genobject.c:221
  33: _PyGen_Send
             at Objects/genobject.c:291
  34: _PyMethodDef_RawFastCallKeywords
             at Objects/call.c:644
  35: _PyMethodDescr_FastCallKeywords
             at Objects/descrobject.c:288
  36: call_function
             at Python/ceval.c:4593
  37: _PyEval_EvalFrameDefault
             at Python/ceval.c:3110
  38: function_code_fastcall
             at Objects/call.c:283
  39: _PyFunction_FastCallKeywords
             at Objects/call.c:408
  40: call_function
             at Python/ceval.c:4616
  41: _PyEval_EvalFrameDefault
             at Python/ceval.c:3124
  42: function_code_fastcall
             at Objects/call.c:283
  43: _PyFunction_FastCallKeywords
             at Objects/call.c:408
  44: call_function
             at Python/ceval.c:4616
  45: _PyEval_EvalFrameDefault
             at Python/ceval.c:3110
  46: PyEval_EvalFrameEx
             at Python/ceval.c:547
  47: _PyEval_EvalCodeWithName
             at Python/ceval.c:3930
  48: _PyFunction_FastCallKeywords
             at Objects/call.c:433
  49: call_function
             at Python/ceval.c:4616
  50: _PyEval_EvalFrameDefault
             at Python/ceval.c:3139
  51: PyEval_EvalFrameEx
             at Python/ceval.c:547
  52: _PyEval_EvalCodeWithName
             at Python/ceval.c:3930
  53: _PyFunction_FastCallKeywords
             at Objects/call.c:433
  54: call_function
             at Python/ceval.c:4616
  55: _PyEval_EvalFrameDefault
             at Python/ceval.c:3110
  56: PyEval_EvalFrameEx
             at Python/ceval.c:547
  57: _PyEval_EvalCodeWithName
             at Python/ceval.c:3930
  58: _PyFunction_FastCallKeywords
             at Objects/call.c:433
  59: call_function
             at Python/ceval.c:4616
  60: _PyEval_EvalFrameDefault
             at Python/ceval.c:3110
  61: function_code_fastcall
             at Objects/call.c:283
  62: _PyFunction_FastCallKeywords
             at Objects/call.c:408
  63: call_function
             at Python/ceval.c:4616
  64: _PyEval_EvalFrameDefault
             at Python/ceval.c:3110
  65: PyEval_EvalFrameEx
             at Python/ceval.c:547
  66: _PyEval_EvalCodeWithName
             at Python/ceval.c:3930
  67: _PyFunction_FastCallDict
             at Objects/call.c:376
  68: _PyObject_FastCallDict
             at Objects/call.c:98
  69: _PyObject_Call_Prepend
             at Objects/call.c:904
  70: PyObject_Call
             at Objects/call.c:245
  71: do_call_core
             at Python/ceval.c:4645
  72: _PyEval_EvalFrameDefault
             at Python/ceval.c:3191
  73: PyEval_EvalFrameEx
             at Python/ceval.c:547
  74: _PyEval_EvalCodeWithName
             at Python/ceval.c:3930
  75: _PyFunction_FastCallKeywords
             at Objects/call.c:433
  76: call_function
             at Python/ceval.c:4616
  77: _PyEval_EvalFrameDefault
             at Python/ceval.c:3124
  78: PyEval_EvalFrameEx
             at Python/ceval.c:547
  79: _PyEval_EvalCodeWithName
             at Python/ceval.c:3930
  80: PyEval_EvalCodeEx
             at Python/ceval.c:3959
  81: PyEval_EvalCode
             at Python/ceval.c:524
  82: run_mod
             at Python/pythonrun.c:1035
  83: PyRun_FileExFlags
             at Python/pythonrun.c:988
  84: PyRun_SimpleFileExFlags
             at Python/pythonrun.c:429
  85: pymain_run_file
             at Modules/main.c:427
  86: pymain_run_filename
             at Modules/main.c:1627
  87: pymain_run_python
             at Modules/main.c:2877
  88: pymain_main
             at Modules/main.c:3038
  89: _Py_UnixMain
             at Modules/main.c:3073
  90: __libc_start_main
  91: _start
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
Aborted
```
They seem to be of size 1 and empty shape, I guess they need special handling
https://stackoverflow.com/questions/54496921/why-is-shape-empty/54497476

**Comment by ijl** [comment]
Fixed in 3.1.2. Thanks for reporting it.

**Comment by sdjespersen** [comment]
Scalar `numpy` arrays still cause an issue, though it is a different issue:
```
Python 3.8.9 (default, Apr 15 2021, 07:37:16)
[Clang 12.0.0 (clang-1200.0.32.29)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import numpy as np
>>> import orjson
>>> np.__version__
'1.18.0'
>>> orjson.__version__
'3.5.4'
>>> orjson.dumps({'a' : np.array(0)}, option=orjson.OPT_SERIALIZE_NUMPY)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: unsupported datatype in numpy array
```
Would you prefer i open a separate issue?
